### ISSUE TITLE
How to use a private packagist to provide an extension?

### ISSUE THREAD

**Issue by macintoshplus** [php/pie#110]
Some companies or state organizations do not allow the servers or the developer's computer to access the Internet. Others have a private PHP extension. I use the `Satis` term to speak of the private packagist.

I try to understand how PIE works with a private composer repository and how to get the DLL to install it. Without reading the code ;-)

I have tried with Satis and xdebug. The installation has noway. 

Composer repository and satis generated files do not contain the PHP extension configuration to build or get the DLL.

For each package information provided by Satis, this is the minimal information:

```json
{
    "name": "smarty/smarty",
    "version": "3.1.7",
    "dist": {
        "url": "https://www.smarty.net/files/Smarty-3.1.7.zip",
        "type": "zip"
    }
}
```

This information is minimal for PHP library. But for PHP extension, some other information is needed: DLL download information, and compilation options...

I have this idea of JSON structure for an extension package

```json
{
    "name": "xdebug/xdebug",
    "version": "3.1.7",
    "sources": {
        "url": "https://xdebug.org/files/xdebug-3.1.7.zip",
        "type": "zip"
        "reference": "tags/3.1.7"
    }
    "dist": {
        "url": "https://www.smarty.net/files/Smarty-3.1.7-8.4-ts-vs17-x86_64.zip",
        "type": "zip"
    },{
        "url": "https://www.smarty.net/files/Smarty-3.1.7-8.4-ts-vs17-x86.zip",
        "type": "zip"
    },{
        "url": "https://www.smarty.net/files/Smarty-3.1.7-8.4-nts-vs17-x86_64.zip",
        "type": "zip"
    },{
        "url": "https://www.smarty.net/files/Smarty-3.1.7-8.4-nts-vs17-x86.zip",
        "type": "zip"
    }
}
```
The source archive contains the composer.json file with the build options and other information. However, no information about the DLLs.

With the public packagist, PIE gets the DLL from the Release Asset and searches the right archive corresponding with the name template (see doc).

How to download DLL from Satis or another private composer repository?

**Comment by macintoshplus** [comment]
@Seldaek, this change has an impact on the composer?

**Comment by Seldaek** [comment]
Yes this would not be doable as you describe just in pie I suppose, unless pie stores/retrieves this stuff in the extra key.

**Comment by macintoshplus** [comment]
> Yes this would not be doable as you describe just in pie I suppose, unless pie stores/retrieves this stuff in the extra key.

Okay, this schema is possible:

```json
{
    "name": "xdebug/xdebug",
    "version": "3.1.7",
    "dist": {
        "url": "https://xdebug.org/files/xdebug-3.1.7.zip",
        "type": "zip"
    }
   "extra": {
        "windows-dist": {
            "url": "https://www.smarty.net/files/Smarty-3.1.7-8.4-ts-vs17-x86_64.zip",
            "type": "zip"
        },{
            "url": "https://www.smarty.net/files/Smarty-3.1.7-8.4-ts-vs17-x86.zip",
            "type": "zip"
        },{
            "url": "https://www.smarty.net/files/Smarty-3.1.7-8.4-nts-vs17-x86_64.zip",
            "type": "zip"
        },{
            "url": "https://www.smarty.net/files/Smarty-3.1.7-8.4-nts-vs17-x86.zip",
            "type": "zip"
        }
    }
}
```

**Comment by asgrim** [comment]
I'm starting to understand what you're asking for now. You are asking for another way of downloading pre-built DLLs (your use case being that you cannot use public internet).

At the moment, the only supported way PIE fetches DLLs is by downloading from the release assets on a GH release: https://github.com/php/pie/blob/main/docs/extension-maintainers.md#windows-support

We can evaluate in the future other ways of downloading prebuilt DLLs such as this, but this isn't something I'll be able to look at just yet.

**Comment by macintoshplus** [comment]
This feature can be used by the maintainer in the `composer.json` found in the version package. In this case, the Packagist must copy the information.

Now you use the GitHub release feature, but some other SVC systems exist and some other source code hosting exists (Zoho, Bitbucket, GitLab...).

**Comment by tillkruss** [comment]
We managed to build a private extension by wrapping it inside a fake extension, would be nice to see 1st party support for this. 

https://github.com/yatsukhnenko/ext-yaler

**Comment by macintoshplus** [comment]
@asgrim The extra data can be put in the `composer.json` file present in the source code archive.

Pros:

* only PIE need change to support this feature

Cons:

* The DLL archive URLs can be changed hardly if changed (format error, moving...).
* Only a new version can be installed with this feature (without private packagist)

**Comment by macintoshplus** [comment]
Hi, I have a POC to 'composer' repository compatible with Pie 1.0.
The goal of my POC is to allow older extensions to install with Pie.

This POC can provide the extension information from the PECL extension information and convert the PECL source archive to a zip archive.

Now, I don't have any way to provide prebuilt DLLs with my private repository.

Any idea how to provide it? Pie extension?

**Comment by macintoshplus** [comment]
Hi @asgrim, I have a POC to 'composer' repository compatible with Pie 1.0.
The goal of my POC is to allow older extensions to install with Pie.

This POC can provide the extension information from the PECL extension information and convert the PECL source archive to a zip archive.

Now, I don't have any way to provide prebuilt DLLs with my private repository.

Any idea how to provide it? Pie extension?

**Comment by macintoshplus** [comment]
Thanks for your answer.

Some PHP "unmaintained" extensions can be built for PHP 8+ without code change. But it's not the subject.

Perhaps, the proxy is a better name. I need to use Pie in a network disconnected from the Internet.

Nexus provides a similar behaviour to dependencies managed with Composer.

I can download the source code archive from the local server after changing the URL in the package metadata.

But no way for the prebuilt DLLs.

_I'm sorry if I expressed myself poorly. Thank you for taking the time to understand my problem._

**Comment by asgrim** [comment]
Hmm, perhaps `\Php\Pie\ComposerIntegration\Listeners\OverrideDownloadUrlInstallListener` is what you want; it reads the `\Php\Pie\Downloading\DownloadUrlMethod` and changes the behaviour. Please see if an additional option there might be suitable and perhaps we can spec something out, depending on what that looks like.

This does feel like a bit of an edge case though; in general PIE (as well as Composer) assume connectivity, so without connectivity, things are somewhat limited.
